### ISSUE TITLE
SQS messages are not received in latest versions

### ISSUE THREAD

**Issue by yairinot-vicarius** [localstack/localstack#3766]
# Type of request: This is a ...

 :ballot_box_with_check:  bug report

# Detailed description

Starting from docker version 0.12.7 including 0.12.8, after using it for a while without problems, suddenly messages to SQS are sent successfully but not received at consumer side.
When using aws cli to get queue attributes, all messages are shown as not visible (all listed by property ApproximateNumberOfMessagesNotVisible)

## Expected behavior

Send and receive SQS messages successfully

## Actual behavior

Messages are not received

## Command used to start LocalStack

Using localstack docker image

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)

Using aws-java-sdk-sqs-1.11.879 

```
awslocal sqs get-queue-attributes --queue-url <queue-url>

{
    "Attributes": {
        "ApproximateNumberOfMessages": "0",
        "ApproximateNumberOfMessagesDelayed": "0",
        "ApproximateNumberOfMessagesNotVisible": "15",
        "CreatedTimestamp": "1616424657.825273",
        "DelaySeconds": "0",
        "LastModifiedTimestamp": "1616424657.825273",
        "MaximumMessageSize": "262144",
        "MessageRetentionPeriod": "345600",
        "QueueArn": "arn:aws:sqs:us-east-1:000000000000:<queue-name>",
        "ReceiveMessageWaitTimeSeconds": "0",
        "VisibilityTimeout": "600"
    }
}

```

**Comment by ldrag** [comment]
I am probably facing same problem. I am using localstack docker image 0.12.9 (but the same happened with 0.12.6).
After sending a message with:
`awslocal sqs send-message --message-body "foo" --queue-url <xyz>`

I can see that the message is available in the queue:
```
awslocal sqs get-queue-attributes --queue-url <xyz>
{
    "Attributes": {
        "ApproximateNumberOfMessages": "1", <-- The message seems to be in the queue
        "ApproximateNumberOfMessagesDelayed": "0",
        "ApproximateNumberOfMessagesNotVisible": "0",
        "CreatedTimestamp": "1617084386.693691",
        "DelaySeconds": "0",
        "LastModifiedTimestamp": "1617084386.693691",
        "MaximumMessageSize": "262144",
        "MessageRetentionPeriod": "60",
        "QueueArn": "arn:aws:sqs:eu-west-1:000000000000:<xyz>",
        "ReceiveMessageWaitTimeSeconds": "0",
        "VisibilityTimeout": "30"
    }
}
```

However an attempt to read the massage
`awslocal sqs receive-message --queue-url <xyz>`
has no results.

Looking at the queue attributes again gives:
```
{
    "Attributes": {
        "ApproximateNumberOfMessages": "0",
        "ApproximateNumberOfMessagesDelayed": "0",
        "ApproximateNumberOfMessagesNotVisible": "1", <-- The message is in flight state
        "CreatedTimestamp": "1617084386.693691",
        "DelaySeconds": "0",
        "LastModifiedTimestamp": "1617084386.693691",
        "MaximumMessageSize": "262144",
        "MessageRetentionPeriod": "60",
        "QueueArn": "arn:aws:sqs:eu-west-1:000000000000:<xyz>",
        "ReceiveMessageWaitTimeSeconds": "0",
        "VisibilityTimeout": "30"
    }
}
```

So to my understanding the message is in flight state, but never reaches the client.

The scenario above does not happen in a deterministic way for me. Occasionally I was able to read a message.

**Comment by Frederick888** [comment]
I've also bumped into this problem... have never seen it under Linux before, but it does happen from time to time under macOS.

Sometimes it happens after my laptop is waken up from sleep but I tested it out and still wasn't able to reproduce it stably, so not sure if this is actually relevant.

**Comment by pinzon** [comment]
Hi, I just made a script to try to force what you're saying. I've been running it for several hours but everything seems to work fine.
Could you try it? [reproducer.zip](https://github.com/localstack/localstack/files/6267595/reproducer.zip)
this are my results:
```
Queue: test-66 Url:http://localhost:4566/000000000000/test-66
Messages Received: 150
Messages Not Received: 0
{
    "Attributes": {
        "ApproximateNumberOfMessages": "111",
        "ApproximateNumberOfMessagesDelayed": "0",
        "ApproximateNumberOfMessagesNotVisible": "39",
        "CreatedTimestamp": "1617743432.153174",
        "DelaySeconds": "0",
        "LastModifiedTimestamp": "1617743432.153174",
        "MaximumMessageSize": "262144",
        "MessageRetentionPeriod": "345600",
        "QueueArn": "arn:aws:sqs:us-east-1:000000000000:test-66",
        "ReceiveMessageWaitTimeSeconds": "0",
        "VisibilityTimeout": "30"
    }
}
```

**Comment by Frederick888** [comment]
@pinzon Tested yours and a slightly modified version (see below). wasn't able to reproduce it... Just turned on Docker experimental features and hopefully I can get a [checkpoint](https://docs.docker.com/engine/reference/commandline/checkpoint/) next time it happens

```sh
#!/usr/bin/env bash

random_id=$((1 + $RANDOM % 100))
name="test-${random_id}.fifo"
url="http://localhost:4566/000000000000/${name}"

aws --endpoint-url=http://h2.lan:4566 sqs create-queue --queue-name $name \
    --attributes '{ "MessageRetentionPeriod": "10800", "ReceiveMessageWaitTimeSeconds": "10", "VisibilityTimeout": "1200", "FifoQueue": "true", "ContentBasedDeduplication": "true" }'

messagesReceived=0
messagesDifferentBody=0
messagesNotReceived=0

for i in {1..150}; do
    aws --endpoint-url=http://h2.lan:4566 sqs send-message --message-body "${i}" --queue-url $url --message-group-id 'foo' >/dev/null
    echo -ne "++Message ${i} sent \033[0K\r"
    sleep 0.3
    message=$(aws --endpoint-url=http://h2.lan:4566 sqs receive-message --queue-url $url --attribute-names All --message-attribute-names All --max-number-of-messages 10)
    if [[ $message ]]; then
        trimmed=$(echo $message | jq -r '.Messages[-1].Body')
        handle=$(echo $message | jq -r '.Messages[-1].ReceiptHandle')
        if [ "$i" == $trimmed ]; then
            messagesDifferentBody=$((messagesDifferentBody + 1))
        fi
        messagesReceived=$((messagesReceived + 1))
        aws --endpoint-url=http://h2.lan:4566 sqs delete-message --queue-url $url --receipt-handle $handle
    else
        messagesNotReceived=$((messagesNotReceived + 1))
    fi
done

echo "-------------------------------------------------"
echo "Queue: ${name} Url: ${url}"
echo "Messages Received: ${messagesReceived}"
# echo "Messages Received with different body: ${messagesDifferentBody}"
echo "Messages Not Received: ${messagesNotReceived}"
aws --endpoint-url=http://h2.lan:4566 sqs get-queue-attributes --queue-url $url
```

**Comment by ldrag** [comment]
@pinzon I was able to reproduce the problem with the script you provided but modified to set a smaller MessageRetentionPeriod. The lower it is the sooner the massages are not received (actually I was even able to set it to a value lower than AWS minimum -> 60). 
An example result for a queue created with retention period 60, ie.

```
-------------------------------------------------
Queue: test-12 Url:http://localhost:4566/000000000000/test-12
Messages Received: 38
Messages Not Received: 112
{
    "Attributes": {
        "ApproximateNumberOfMessages": "130",
        "ApproximateNumberOfMessagesDelayed": "0",
        "ApproximateNumberOfMessagesNotVisible": "20",
        "CreatedTimestamp": "1617790306.223556",
        "DelaySeconds": "0",
        "LastModifiedTimestamp": "1617790306.223556",
        "MaximumMessageSize": "262144",
        "MessageRetentionPeriod": "60",
        "QueueArn": "arn:aws:sqs:eu-west-1:000000000000:test-12",
        "ReceiveMessageWaitTimeSeconds": "0",
        "VisibilityTimeout": "30"
	}		
}	
```

**Comment by pinzon** [comment]
@ldrag could you provide more info in your setup? I'm still unable to reproduce even with the lower retention period.

```

➜  3766 bash reproduce.sh
{
    "QueueUrl": "http://localhost:4566/000000000000/test-85.fifo"
}
-------------------------------------------------
Queue: test-85.fifo Url: http://localhost:4566/000000000000/test-85.fifo
Messages Received: 150
Messages Not Received: 0
{
    "Attributes": {
        "ApproximateNumberOfMessages": "0",
        "ApproximateNumberOfMessagesDelayed": "0",
        "ApproximateNumberOfMessagesNotVisible": "0",
        "CreatedTimestamp": "1617897065.229335",
        "DelaySeconds": "0",
        "LastModifiedTimestamp": "1617897065.229335",
        "MaximumMessageSize": "262144",
        "MessageRetentionPeriod": "1",
        "QueueArn": "arn:aws:sqs:us-east-1:000000000000:test-85.fifo",
        "ReceiveMessageWaitTimeSeconds": "1",
        "VisibilityTimeout": "1",
        "FifoQueue": "true",
        "ContentBasedDeduplication": "true"
    }
}
```

**Comment by Frederick888** [comment]
I can also reproduce the issue using my script with message retention set to 30:

```
{
    "QueueUrl": "http://h2.lan:4566/000000000000/test-59.fifo"
}
-------------------------------------------------
Queue: test-59.fifo Url: http://localhost:4566/000000000000/test-59.fifo
Messages Received: 15
Messages Not Received: 135
{
    "Attributes": {
        "ApproximateNumberOfMessages": "30",
        "ApproximateNumberOfMessagesDelayed": "0",
        "ApproximateNumberOfMessagesNotVisible": "105",
        "CreatedTimestamp": "1617926338.851645",
        "DelaySeconds": "0",
        "LastModifiedTimestamp": "1617926338.851645",
        "MaximumMessageSize": "262144",
        "MessageRetentionPeriod": "30",
        "QueueArn": "arn:aws:sqs:us-east-1:000000000000:test-59.fifo",
        "ReceiveMessageWaitTimeSeconds": "10",
        "VisibilityTimeout": "1200",
        "FifoQueue": "true",
        "ContentBasedDeduplication": "true"
    }
}                            
```

**Comment by ldrag** [comment]
I am starting localstack as one of the services in my docker-compose. The interesting part looks like this:
```
localstack:
    image: localstack/localstack
    environment:
      - SERVICES=sns,sqs,sts
      - DEFAULT_REGION=eu-west-1
    volumes:
      - ./foo/bar:/docker-entrypoint-initaws.d
    expose:
      - "4566"
```
The container output reveals I am running version 0.12.9:
```
Waiting for all LocalStack services to be ready
Starting local dev environment. CTRL-C to quit.
LocalStack version: 0.12.9
...
LocalStack build date: 2021-03-29
LocalStack build git hash: 16d414f
Starting edge router (https port 4566)...
Starting mock STS service on http port 4566 ...
Starting mock SNS service on http port 4566 ...
Starting mock SQS service on http port 4566 ...
```

I also have a simple init script in `docker-entrypoint-initaws.d` volume:
```
topicArn=$(awslocal sns create-topic --name some-topic | grep -oE 'arn:[^"]+')
queueUrl=$(awslocal sqs create-queue --queue-name some-queue --attributes MessageRetentionPeriod=60 | grep -oE 'http[^"]+')
awslocal sns subscribe --topic-arn $topicArn --protocol sqs --notification-endpoint $queueUrl
```

I hope this will be helpful for you.

**Comment by fhsgoncalves** [comment]
We are also experiencing this behavior on version 0.12.9, but we also tested against versions 0.11.6 and 0.10.9 and the issue was still occurring.

Interesting is that it only occurs when I have 4 or 5 processes receiving messages from different queues locally, but with just one process and one queue, everything is ok.

The workaround I'm using for now is to set `SQS_PROVIDER` env var to `elasticmq`, so elasticmq is used instead of `moto`. I hope it helps in the analysis, to compare moto with elasticmq implementations.

**Comment by pinzon** [comment]
@ldrag thank you for sharing your setup. When I run localstack inside a docker container and subscribe the queue , some messages stop being received.
So I'm trying to fix it.

**Comment by dittos** [comment]
Is this related to https://github.com/spulec/moto/pull/3924 ?

If I create a queue with very short `MessageRetentionPeriod` (60 secs.), wait 60 seconds, then all messages sent after are not received.

**Comment by pinzon** [comment]
I'll take a look and triage again, since I was able to reproduced only in a very specific way regardless of the queue settings.

I found a solution that works for me and it's to redefine the original function that handles the get_messages request but it does seem right.

There has been some chat about going back to the esasticmq for backend as default. for the moment If don't find a better solution I guess I'll make a PR with it.

**Comment by pinzon** [comment]
Hi everyone, I just took another look into triaging this issue, after making sure that my moto lib was updated and testing against AWS, seems to me that the issue has been fixed thanks to spulec/moto#3924 .
I'm going to close this issue. Please feel free to comment if this still happens to you.
